automation/nodejs/library.mjs

```diff
@@ -0,0 +1,187 @@
+import path from 'path'
+import simpleGit from 'simple-git'
+import { statSync, rmSync } from 'fs'
+import { copyFile } from 'fs/promises'
+import projectsList from '../../projects.list.mjs'
+
+export const data = {
+  run_date: (d => `${d.getFullYear()}${d.getMonth()<9?'0':''}${d.getMonth()+1}${d.getDate()<10?'0':''}${d.getDate()}`)(new Date()),
+}
+
+export const paths = {
+  project_root: projectsList.config.project_root,
+  working_root: process.env?.WORKING_ROOT ?? path.join(projectsList.config.project_root, '_working_'),
+}
+
+export const projects = projectsList.projects
+
+export const preseeds = projectsList.preseeds
+
+/**
+ * If the `working_root` folder exists, remove it.
+ */
+export function clearWorkingRoot () {
+  if (fileExists(paths.working_root)) {
+    console.info('removing the working_root:', paths.working_root)
+    rmSync(paths.working_root, { force: true, recursive: true })
+  }
+}
+
+/**
+ * At the target directory, create a git repo fetch only the given ref from the given
+ * repo and check it out as branch `target`.
+ *
+ * @param {string} dir Target working directory
+ * @param {string} repoUrl Git URL for the remote repository to fetch from
+ * @param {string} repoRef Remote ref to fetch, defaults to `master`
+ * @param {string} localBranch Name of the local branch for the `remoteRef`, defaults to `target`
+ *
+ * @returns {string} Full hash code of the HEAD commit
+ */
+export async function checkoutRepo (dir, repoUrl, repoRef = 'master', localBranch = 'target') {
+  const git = simpleGit(dir)
+  const res =
+    await git
+      .init()
+      .addRemote('origin', repoUrl)
+      .fetch([
+        '--no-tags',
+        '--prune',
+        '--depth=1',
+        'origin',
+        `${repoRef}:${localBranch}`,
+      ])
+      .checkout(localBranch)
+      .log()
+
+  return res.latest.hash
+}
+
+/**
+ * Check if a file exists at the given path.
+ */
+export function fileExists (path) {
+  const stat = statSync(path, { throwIfNoEntry: false })
+  if (stat) {
+    return true
+  } else {
+    return false
+  }
+}
+
+/**
+ * Checkout a remote ref and verify repo contents.
+ *
+ * @param {string} targetDir Target working directory
+ * @param {string} name Working name of the repo
+ * @param {Object} info Info about the repository
+ * @param {string} info.git_url git clone URL
+ * @param {string} info.branch remote branch to fetch and verify
+ * @param {string} info.folder where to look for necessary files, defaults to `/`
+ *
+ * @returns {Object} result Verification result
+ * @returns {string} result.commitId The head commit id for the remote branch
+ * @returns {string[]} result.importantFiles Set of important file verified to exist in the repo
+ */
+export async function checkoutAndVerifyRepo(targetDir, name, { git_url, branch, folder = '/' }) {
+  let commitId
+  try {
+    commitId = await checkoutRepo(targetDir, git_url, branch)
```

Comment:
> 1. we could get rid of `let` with  `const commitId = await checkoutRepo(...).catch(e => ...)`

That seems a bit more confusing than not ... blending the try/catch/await style with the Promise callback style

> 
>     2. the original call stack might get lost after wrapping and throwing new exception - if we care we can log and re-throw original exception

Turns out there is a second Error constructor that takes a `cause`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#parameters

Reply:
> That seems a bit more confusing than not ... blending the try/catch/await style with the Promise callback style
>

Yep. However async style shows some weak sides here - enforcing `let`. Alternatively we could extract the call to a wrapper i.e.` const commitId = await checkoutRepoWrapper()`

> Turns out there is a second Error constructor
>

+1

Reply:
> > That seems a bit more confusing than not ... blending the try/catch/await style with the Promise callback style
> 
> Yep. However async style shows some weak sides here - enforcing `let`. Alternatively we could extract the call to a wrapper i.e.` const commitId = await checkoutRepoWrapper()`
>

1. It is just for a script really, so doesn't need to be "best"
2. A Java style try-with-resource construct would be nicer, but we don't have it in JS
3. I don't see a problem with "enforcing `let`" as it reads just fine
4. I tried this but it doesn't work so I'm not even sure your suggestion is possible:

```js
async function doWorkStep () {
  return 'some work'
}

async function doWork () {
  const step = await doWorkStep.catch(err => { throw new Error('error in doWorkStep', {cause:err}) })
  console.log('work is done, step:', step)
}

try {
  await doWork()
} catch (e) {
  console.error(e)
}
```

Reply:
> 1. It is just for a script really, so doesn't need to be "best"
>
True. I'm OK with leaving it. 

> 4. I tried this but it doesn't work so I'm not even sure your suggestion is possible:
>

There is a typo. You need `await doWorkStep().catch(err =>...)`.
Some alternatives:
1. `const [err, data] = await to(doWorkStep)`  - `to()` is a simple wrapper from [await-to-js](https://www.npmjs.com/package/await-to-js) 
2. custom wrapper using `return await`
```
async function wrapper() {
  try {
    return await doWork()
  } catch(e) {
    throw new Error()
  }
}
```

Reply:
Standard JS async/await/try/catch for the errors instead of extra wrapper code is easiest at this point, even if the `let` declaration hurts eyes. ;-)